lib/elixir/lib/kernel.ex

```diff
@@ -1592,43 +1592,32 @@ defmodule Kernel do
 
   """
   defmacro left <> right do
-    concats = extract_concatenations({:<>, [], [left, right]})
+    concats = extract_concatenations({:<>, [], [left, right]}, __CALLER__)
     quote(do: <<unquote_splicing(concats)>>)
   end
 
   # Extracts concatenations in order to optimize many
   # concatenations into one single clause.
-  defp extract_concatenations({:<>, _, [left, right]}) do
-    [wrap_concatenation(left, :left) | extract_concatenations(right)]
+  defp extract_concatenations({:<>, _, [left, right]}, caller) do
+    [wrap_concatenation(left, :left, caller) | extract_concatenations(right, caller)]
   end
 
-  defp extract_concatenations(other) do
-    [wrap_concatenation(other, :right)]
+  defp extract_concatenations(other, caller) do
+    [wrap_concatenation(other, :right, caller)]
   end
 
-  defp wrap_concatenation(binary, _side) when is_binary(binary) do
+  defp wrap_concatenation(binary, _side, _caller) when is_binary(binary) do
     binary
   end
 
-  defp wrap_concatenation(other, side) do
-    {:::, [],
-     [
-       {:expand_concatenation_argument, [context: Elixir, import: Kernel], [other, side]},
-       {:binary, [], nil}
-     ]}
-  end
-
-  @doc false
-  # Used internally to check the arguments expansion of <> operator.
-  # This function is private and must be used only internally.
-  defmacro expand_concatenation_argument(other, side) do
+  defp wrap_concatenation(other, side, caller) do
     expand =
-      case bootstrapped?(Macro) do
-        true -> &Macro.expand(&1, __CALLER__)
-        false -> & &1
+      case caller.context do
```

Comment:
We should still check if we have `bootstrapped?(Macro)` so we don't need to change other parts of Elixir codebase. Otherwise it is a very common pitfall to run into. :) We just skip the check altogether if not bootstrapped yet.

Reply:
Basically, if `bootstrapped?(Macro) and caller.context == :match` is what we are going after here.

Another way to structure this code, which would probably be better, would be to move the `match` checks and expansion to a separate function so we don't have a large `case`. So basically we check if it is a list, atom, etc, first then we call a function, only if match, that does the expansion (if available) and run the remaining checks.

Reply:
One thing that I was thinking was to always expand to check the errors, but add to the final expression the original value. It will expand two times, but it won't cause any issues with the env and it won't allow any strange error when checking the generated expression